components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/JWTAccessTokenOIDCClaimsHandler.java

```diff
@@ -0,0 +1,279 @@
+                }
+            }
+        }
+        return jwtClaimsSetBuilder.build();
+    }
+
+    /**
+     * Check weather claim value if multi valued or not.
+     *
+     * @param claimKey Claim key.
+     * @param claimValue Claim value.
+     * @param multiAttributeSeparator Multi attribute separator.
+     * @return True if claim value is multi valued, false otherwise.
+     */
+    private boolean isMultiValuedAttribute(String claimKey, String claimValue, String multiAttributeSeparator) {
+
+        // Address claim contains multi attribute separator but its not a multi valued attribute.
+        if (claimKey.equals(ADDRESS)) {
+            return false;
+        }
+        // To format the groups claim to always return as an array, we should consider single group as
+        // multi value attribute.
+        if (claimKey.equals(GROUPS)) {
+            return true;
+        }
+        return StringUtils.contains(claimValue, multiAttributeSeparator);
+    }
+
+    /**
+     * Get user roles.
+     *
+     * @param authenticatedUser Authenticated user.
+     * @param clientId Client Id.
+     * @return Array of user roles.
+     */
+    private String[] getUserRoles(AuthenticatedUser authenticatedUser, String clientId)
+            throws IdentityOAuth2Exception {
+
+        String applicationId = getApplicationId(clientId, authenticatedUser.getTenantDomain());
+        List<String> userRoleIds = AuthzUtil.getUserRoles(authenticatedUser, applicationId);
+        List<RoleV2> rolesAssociatedWithApp = getRolesAssociatedWithApplication(applicationId,
+                authenticatedUser.getTenantDomain());
+        return rolesAssociatedWithApp.stream().filter(role -> userRoleIds.contains(role.getId())).map(RoleV2::getName)
+                .map(this::appendInternalDomain).toArray(String[]::new);
+    }
+
+    /**
+     * Get roles associated with the application.
+     *
+     * @param applicationId Application Id.
+     * @param tenantDomain Tenant Domain.
+     * @return List of roles associated with the application.
+     */
+    private List<RoleV2> getRolesAssociatedWithApplication(String applicationId, String tenantDomain)
+            throws IdentityOAuth2Exception {
+
+        ApplicationManagementService applicationMgtService = OAuth2ServiceComponentHolder.getApplicationMgtService();
+        try {
+            return applicationMgtService.getAssociatedRolesOfApplication(applicationId, tenantDomain);
+        } catch (IdentityApplicationManagementException e) {
+            throw new IdentityOAuth2Exception("Error while getting roles associated with application : " + applicationId
+                    + " tenant : " +  tenantDomain);
+        }
+    }
+
+    /**
+     * Get application Id.
+     *
+     * @param clientId Client Id.
+     * @param tenantDomain Tenant Domain.
+     * @return Application Id.
+     */
+    private String getApplicationId(String clientId, String tenantDomain) throws IdentityOAuth2Exception {
```

Comment:
aren't there a existing util method for this?

Reply:
no